yjit/src/codegen.rs

```diff
@@ -2775,15 +2775,23 @@ fn gen_setinstancevariable(
             Counter::setivar_megamorphic,
         );
 
-        asm.spill_temps(); // for ccall (must be done before write_val is popped)
         let write_val;
 
+        if !stack_type.is_imm() {
+            asm.spill_temps(); // for ccall (must be done before write_val is popped)
+        }
+
         match ivar_index {
             // If we don't have an instance variable index, then we need to
             // transition out of the current shape.
             None => {
                 let (new_shape_id, needs_extension, ivar_index) = new_shape.unwrap();
                 if let Some((current_capacity, new_capacity)) = needs_extension {
+                    // We already spilled temps in the case the stack type is _not_ an immediate.
+                    // If it is an immediate, but we need to expand the object, then spill temps
+                    if stack_type.is_imm() {
+                        asm.spill_temps(); // for ccall (must be done before write_val is popped)
+                    }
```

Comment:
```suggestion
        match ivar_index {
            // If we don't have an instance variable index, then we need to
            // transition out of the current shape.
            None => {
                let (new_shape_id, needs_extension, ivar_index) = new_shape.unwrap();
                if let Some((current_capacity, new_capacity)) = needs_extension {
```

question: What happens if we just remove `spill_temps()`? Does it crash if you just let `asm.ccall` spill them?

Reply:
Good question, I haven't tried. I didn't think about `ccall` automatically spilling as I wrote the patch. I'll check it out!

Reply:
@k0kubun ya, I agree this would have a problem in combination with #7588. But I don't think we're sure if we want to merge that PR? I _think_ this PR is a win regardless of #7588

Reply:
I shouldn't have mentioned https://github.com/ruby/ruby/pull/7588 since that confused you, but the problem (the `skip_wb` branch) exists whether you merge it or not. Unless you remove a conditional `asm.ccall` guarded by `skip_wb` label, you still have to spill them on L2827.